### DISCUSSION TITLE
USB MSC, Copy Files Larger Than Available RAM Size?

### DISCUSSION THREAD

**Discussion by stanleyshly** [hathach/tinyusb#1294] [Q&A] [answered]
Looking at the cdc_msc example in the msc_disk.c file, this section of code exists:
`  if ( lba >= DISK_BLOCK_NUM ) return -1;
`
This seems to suggest, that for both read and write, if the data that is trying to be copied is larger than the remaining ram, it is not possible. However, if I'm using a filesystem that can read the data in blocks that each individually fit into memory, is there anyway to  read and write files larger than the size of RAM?

**Comment by Ryzee119** [comment, marked as answer]
No such limit really exists, thats just the example code which uses a 'RAM disk'.
Just implement tud_msc_write10_cb and tud_msc_read10_cb accordingly and redirect the reads/writes to your filesystem. These calls are in sector sizes (nominally only 512 bytes)

**Reply by stanleyshly** [reply]
I see, sorry for the misunderstanding, I'm not very good with microcontrollers yet, this is my first major foray embedded systems in general. 
Just clarify, does that mean, if I'm using littlefs as the backing FS, I can do something like and have everything work properly?
```
lfs_file_read(&lfs, &file, &buffer, size)
memcpy(buffer, addr, bufsize);
```
However, do I need a loop to copy in 512 bytes chunks?

**Reply by Ryzee119** [reply]
MSC is more suited to direct access to the storage medium and the host OS can format and manage the whole memory area.

If you use lfs would need a file that represents your usb MSC (i.e. a 8mb filenamed msc.bin or something) where you would use lfs to read/write to that file. Dont forget to lfs_seek to the correct part of the file (lba*512).

littlefs in the backend would use low level read/write commands. These may be better suited for MSC.

**Reply by stanleyshly** [reply]
Would you be willing to clarify what you mean low level read/write? I'm not sure how they would be better.
That being said, having a msc example with littlefs would be super nice, I'm sure many project would benefit from msc capabilities while still being able to write/read/create and delete files from a drive. 
But, you have answer my question, so I'll mark your first one as the answer, and once again, thank you for your help.
I'll start a new discussion/feature request instead.
